### ISSUE TITLE
documentation, what is missing ?

### ISSUE THREAD

**Issue by leonoel** [leonoel/missionary#41]

**Comment by ribelo** [comment]
The documentation is good, but only for someone who catches the concept.

However, for someone who feels reasonably proficient in core.async, it's like colliding with a wall. Everything is different, and it's a little hard to translate things, to missionary.

I don't have a problem with reading tests, but in this case they are not always readable either, especially since you first have to understand how the underlying macros work.

IMHO something like this would suffice, touching on all the features available and how to use them.
https://github.com/clojure/core.async/blob/master/examples/walkthrough.clj

**Comment by martinklepsch** [comment]
Some ideas: 

- A walkthrough of some common workflows with ClojureScript, e.g. subscribe to DOM events, start/end flows etc.
- A list of the various macros `m/?` `m/ap` and so on with descriptions of what they do, CLJ/S compatibility, and a "how to read them"

> IMHO something like this would suffice, touching on all the features available and how to use them.
https://github.com/clojure/core.async/blob/master/examples/walkthrough.clj

There are some tools (like [transcriptor](https://github.com/cognitect-labs/transcriptor)) to turn REPL sessions into tests, maybe that'd be a good format for this.

**Comment by steffan-westcott** [comment]
Regarding the overall structure of documentation for any software project, I've come across this pattern which has been applied to several high profile projects: https://documentation.divio.com/ (The video is worth watching). I believe adopting this pattern would help both beginners and experienced users well.

**Comment by xificurC** [comment]
AFAIK the creator of that documentation structure moved it to https://diataxis.fr/

**Comment by xificurC** [comment]
A cookbook (called how-to guides above) is what I miss the most. The current 3 guides are nice, but more, shorter examples of common patterns would help me personally.

@leonoel any reason you're asking for suggestions? Did someone complain or do you find the documentation lacking?

**Comment by leonoel** [comment]
> any reason you're asking for suggestions? Did someone complain or do you find the documentation lacking?

Both. A significant part of the library is not covered at all by current docs, several users have asked for more resources, and the learning curve feels steep. Also I'd like to understand better what newcomers struggle with.

**Comment by xificurC** [comment]
I still struggle understanding how will my code run, most importantly when will it block and what to do when I don't want a task to block. I think it's typically with tasks, flows are easier on this front.

Some of the primitives (e.g. in happy eyeballs `race` of `dfv`) are not part of the documentation, which makes happy eyeballs harder to understand.

How about we start with a couple of wiki pages? Specter has [this nice list of navigators](https://github.com/redplanetlabs/specter/wiki/List-of-Navigators) that I visit often and find valuable and short explanations with simple examples. A second page could be a cookbook of patterns.

**Comment by bsless** [comment]
This partially relates to documentation and partially to naming, but I find that runes and acronyms make it harder to grok the code, examples and documentation. Until I'm fluent, I have to keep reminding myself, `sp` is sequential process, `ap` is asyncrhonous-process, not to mention all the variations on pulling and forking. If they had clearer names I think it would go a long way towards making the code more approachable to new arrivals.
There are also some fundamental concepts such as forking which could be clarified.
Also, for people who learn more by following things up from first principles than by example, it could be beneficial to create a continuity between the task and flow specification to the primitives in the library.
This issue might be partially related, too https://github.com/leonoel/flow/issues/1

**Comment by mjmeintjes** [comment]
I think it would be useful to have more examples of how to integrate missionary with existing libraries and tools (in addition to the current documentation on integrating with RxJava):

Some examples:
- Converting core.async channel to flow
- Converting promise to task
- Consuming a BlockingQueue
- CompletableFuture to task 
- File IO
- HTTP

It might be worth adding a wiki or something similar for managing and working on these types of snippets and examples.

**Comment by bsless** [comment]
Trying to see if I've been paying attention in class, CompletableFuture to task should look something like:
```clojure
(import '(java.util.concurrent Executor CompletableFuture)
        '(missionary.impl Thunk))

(defn cf->task
  [^CompletableFuture cf]
  (fn [success failure]
    (.handleAsync
     cf
     (reify java.util.function.BiFunction
       (apply [_ r e]
         (if (instance? Exception e)
           (failure e)
           (success r))))
     Thunk/cpu)
    (fn [] (.cancel cf true))))

(let [cf (CompletableFuture.)
      t (cf->task cf)
      success! (fn [res] (println 'yay! res))
      fail! (fn [e] (println 'Error! (ex-message e)))]
  (t success! fail!)
  (.complete cf 2))


(let [cf (CompletableFuture.)
      t (cf->task cf)
      success! (fn [res] (println 'yay! res))
      fail! (fn [e] (println 'Error! (ex-message e)))]
  (t success! fail!)
  (.completeExceptionally cf (Exception. "failed!")))
```

**Comment by dspiteself** [comment]
A great thing to add to the documentation is `what is functional effect and streaming system?` `why use functional effects?`
My answer would be something like below. It could be on separate page or inlined at the top of the main page.

A side effect is when a function relies on, or modifies, something outside its parameters to do something. 
For example: `(println "Hello world!")`
A functional effect is value that describes an effect, without actually executing it. In missionary we represent function effects using tasks. 
For example creating a sequential process  for the effect hello world `(def hello-world-task (sp (println "Hello world!")))`
now will not print to the console until the task is ran
`(m/? hello-world-task)`
will run the task `hello-world-task` which will print "Hello world!" to the console.
Using functional effects yields many benefits including improvement to testability and composability.

**Comment by kawas44** [comment]
Hi,

I have added a [walkthrough][] "à la core.async" focused on Missionary Clojure implementation. The last example about flow is an implementation of the example from the core.async walkthrough using `alt!!`.

I have added a [cheatsheet][] "à la specter", for the moment only about tasks and flows (I dont understand yet how to properly use `dfv`, `mbx`, `sem`, etc)

[walkthrough]: https://github.com/leonoel/missionary/wiki/Basic-Walkthrough:-Tasks-&-Flows
[cheatsheet]: https://github.com/leonoel/missionary/wiki/Cheatsheet

**Comment by xificurC** [comment]
@leonoel

> I have a WIP tutorial on the reactor

could you share it even in its raw state?

**Comment by leonoel** [comment]
@kawas44 Thank you, these resources are highly valuable. I have some remarks about the walkthrough :
* `(m/ap (for [i (range 4)] (m/?> i)))` this is invalid code (because `i` is not a flow, and because `m/?>` is not allowed in closures)
* in the last example, the `ap` block is actually fully synchronous and single threaded. We could make it async using `via`, the execution would be closer to what happens with core.async. 
```clojure
(let [begin (System/currentTimeMillis)
      ;; create a flow of values generated by asynchronous tasks
      inputs (repeat 1000 (m/via m/cpu "hi")) ;; a task has no identity, it can be reused
      values (m/ap
               (let [flow (m/seed inputs)     ;; create a flow of tasks to execute
                     task (m/?= flow)]        ;; from here, fork on every task in **parallel**
                 (m/? task)))                 ;; get a forked task value when available

      ;; drain the flow of values and count them
      n (m/? ;; tasks are executed, and flow is consume here!!
         (m/reduce (fn [acc v]
                     (assert (= "hi" v))
                     (inc acc))
                   0 values))]
  (println "Read" n "msgs in" (- (System/currentTimeMillis) begin) "ms"))
```

**Comment by kawas44** [comment]
> * `(m/ap (for [i (range 4)] (m/?> i)))` this is invalid code (because `i` is not a flow, and because `m/?>` is not allowed in closures)

Oups, didn't try to reduce it :grimacing: 
I can fix it with `amb>` which I understood while writing the cheatsheet :)
```clojure
(m/ap (m/amb> 1 2 3))
```

> * in the last example, the `ap` block is actually fully synchronous and single threaded. We could make it async using `via`, the execution would be closer to what happens with core.async.

Yes I knew it run on the same thread. I am still a little bit lost on the vocabulary. Maybe "concurrent" would have been a better word. Anyway, if you dont mind, I will  copy/paste your version.

**Comment by dustingetz** [comment]
IMO any/all tutorials need to work in ClojureScript (so m/? outside process block should not be in the tutorial, instead describe the task interface and how to run it directly

also see https://github.com/leonoel/task

```
(def task (fn [success! failure!] (success! 1) (fn cancel! [])))
(def cancel! (task #(prn ::success %) #(prn ::failure %)))
```

**Comment by leonoel** [comment]
@dustingetz makes sense to me, it also prevents a potential confusion between parking `m/?` and blocking `m/?`.

**Comment by xificurC** [comment]
[hello task](https://cljdoc.org/d/missionary/missionary/b.26/doc/readme/tutorials/hello-task#sequential-composition) suffers the same issue btw. It even mentions the task works in cljs (which it does), but the blocking `m/?` call is misleading there too

**Comment by leonoel** [comment]
@xificurC 
> could you share it even in its raw state?

I'd like to make a step-by-step guide to solve the petrol pump problem popularized by the sodiumFRP team.

My current [implementation](https://gist.github.com/leonoel/c5e32b65ec7b6ab4b5b45772082a3d85) relies on unreleased features, you'll need to compile the master branch locally.

**Comment by dustingetz** [comment]
```
(tests
  "Missionary signals"
  (def !x (atom 0))                                         ; atoms model variable inputs
  (def >x (m/watch !x))                                     ; "recipe" for a signal derived from atom

  ; a signal is a "continuous flow" in Missionary jargon
  ; signals (flows) are recipes, like Haskell IO actions.
  ; Like IO actions, they are pure function values (thunks)
  ; and do not perform side effects until you run them.
  (fn? >x) := true                                          ; thunk (implementation detail)
  ; The atom has not been subscribed to yet, because >x is a pure value

  ; Flow thunks concretely have the structure (fn [notify! terminate!] !iterator),
  ; see https://github.com/leonoel/flow#specification
  (def !it (>x (fn [] (! ::notify))
               (fn [] (! ::terminate))))
  % := ::notify                                             ; lazy flow is ready to be sampled
  @!it := 0                                                 ; sample
  (swap! !x inc)                                            ; trigger a change
  % := ::notify                                             ; flow is ready again
  @!it := 1                                                 ; sample
  )
```

**Comment by PEZ** [comment]
@dustingetz `!` in those tests does not resolve for me. Do I need to refer something? `m/!` expects zero arguments...

**Comment by PEZ** [comment]
I specifically started to browse the various resources to try to figure out if I can compose flows. I have a continuous flow that I want to build another continuous flow from.

I then quite quickly realized I needed to start from first principles instead. Though I think I need a combination of the concepts laid out and walkthroughs that I can run in the REPL. I like how the Tasks & Flow walkthrough starts. 😄

**Comment by dustingetz** [comment]
`!` there is `hyperfiddle.rcf/tap`, which used to be `hyperfiddle.rcf/!`

**Comment by leonoel** [comment]
@awwx Thank you, this is a valuable feedback.

It is allowed to spawn the same effect (task or flow) more than once, it will result in performing the actions described by this effect more than once. A publisher is a memoized view of an effect, it is also an effect and the action it performs is called a subscription. Running a publisher twice registers two subscriptions, but the underlying effect is spawned only once due to memoization.

I acknowledge the wording confusion. Please note :
* the wiki is outdated. `reactor`, `signal!` and `stream!`  are deprecated and superseded by `signal` and `stream` (without bang suffix)
* `publish` and `subscribe` are about reactive streams interop and orthogonal to missionary publishers. The name collision is unfortunate, but the interface `Publisher` in reactive streams is not necessarily about publishing as it's merely an effect descriptor (like flow)
* the documentation is currently being rewritten, there will be a detailed explanation about the effect protocols and the new publisher design, I will make sure the terminology is consistent and up to date.

**Comment by awwx** [comment]
ok, as an introduction in the documentation, perhaps something like the following (please feel free to use any of this in the documentation, if it's correct and useful):

An instance of a flow transfers a series of values from a single producer to a single consumer.  (Flows can be multiplexed if needed by using a Missionary publisher).

If computers were infinitely fast, a consumer would always be able to keep up with its publisher.  Since sometimes a consumer might not able to keep up, we need a strategy for what to do when it doesn't.  Two common strategies are:

- We have a flow of discrete values, such as events, where it's important to process every value (and we hope that the consumer falling behind is only temporary).  Then either the values need to only be generated when the consumer is ready to receive them (back pressure), or, if the values have already been generated, they need to be retained so that they can be delivered when the consumer is ready.

- We have a flow for a continuous signal, where when the consumer is ready to receive a new value, the consumer only cares about the most recent value of the signal.  If the consumer has fallen behind, old values can be discarded.

For example, `m/seed` returns a discrete flow from a Clojure collection, and each item in the collection will be delivered to the consumer when the consumer is ready to receive it, regardless of how long the consumer takes to receive the values.

`m/watch` returns a continuous flow reflecting the current value of a Clojure reference type.  As long as the consumer can keep up, each update of the reference type will be delivered, but, if the consumer falls behind, when the consumer is ready to receive again it will get the latest value of the reference type without having seen intermediate values.

Both scenarios are accommodated in the flow specification by using a handshake protocol.  The consumer notifies the producer when its ready to receive a new value.  Neither the producer nor the consumer gets ahead of the other: the producer waits until it's been notified by the consumer before sending the consumer a new value, and the consumer waits until it's received a value before notifying the producer again.

**Comment by awb99** [comment]
I suggest that **depreciated functions/macros are NOT included in the cljdocs**.

This would make sense because once something gets depreciated, then new users (who would read the clj-docs) should not use it, so it should not be in the clj-docs.

This is important because missionary has a lot of operators and it is not easy to get ones head around to find the right
one. If the cljdoc list gets reduced a little, then this helps.

Thanks!

**Comment by respatialized** [comment]
If possible, I think the docs could use a simple "fullstack" example that shows how to use Missionary over the network to convey value streams and signals between the JVM and ClojureScript runtimes.
